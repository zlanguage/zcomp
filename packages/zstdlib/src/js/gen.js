"use strict";

const $Z = require("@zlanguage/zstdlib");
const matcher = require("@zlanguage/zstdlib/src/js/matcher");

const $eq = $Z.$eq;
const isObject = $Z.isObject;
const typeOf = $Z.typeOf;
const stone = $Z.stone;
const log = $Z.log;
const copy = $Z.copy;
const assertBool = $Z.assertBool;
const $plus = $Z.$plus;
const $minus = $Z.$minus;
const $star = $Z.$star;
const $slash = $Z.$slash;
const $percent = $Z.$percent;
const $carot = $Z.$carot;
const pow = $Z.pow;
const $lt = $Z.$lt;
const $gt$eq = $Z.$gt$eq;
const $gt = $Z.$gt;
const $lt$eq = $Z.$lt$eq;
const not = $Z.not;
const $plus$plus = $Z.$plus$plus;
const m = $Z.m;
const both = $Z.both;
const either = $Z.either;
const and = $Z.and;
const or = $Z.or;
const JS = $Z.JS;
const assertType = $Z.assertType;
const typeGeneric = $Z.typeGeneric;
const chan = $Z.chan;
const send = $Z.send;
const to = $Z.to;
const til = $Z.til;
const by = $Z.by;
const curry = $Z.curry;
const $or$gt = $Z.$or$gt;
const $gt$gt = $Z.$gt$gt;
const $lt$lt = $Z.$lt$lt;
const handleErr = $Z.handleErr;

const constant = function (value) {
  return function () {
    return value;
  };
};
const integer = function (from = 0, to = Number["MAX_SAFE_INTEGER"], step = 1) {
  return function () {
    if (assertBool($lt(from, to))) {
      const result = from;
      from = $plus(from, step);
      return result;
    }
  };
};
const element = function (array, gen = integer(0, array["length"])) {
  return function (...args) {
    const elementNumber = gen(...args);
    if (assertBool(not($eq(elementNumber, undefined)))) {
      return array[elementNumber];
    }
  };
};
const property = function (object, gen = element(Object["keys"](object))) {
  return function (...args) {
    const key = gen(...args);
    if (assertBool(not($eq(key, undefined)))) {
      return [key, object[key]];
    }
  };
};
const collect = function (gen, arr) {
  return function (...args) {
    const val = gen(...args);
    if (assertBool(not($eq(val, undefined)))) {
      arr["push"](val);
    }
    return val;
  };
};
const repeat = function (gen) {
  while (true) {
    if (assertBool($eq(gen(), undefined))) {
      break;
    }
  }
};
const harvest = function (gen) {
  const arr = [];
  repeat(collect(gen, arr));
  return arr;
};
const limit = function (gen, count = 1) {
  return function (...args) {
    if (assertBool($gt$eq(count, 1))) {
      $minus$eq(count, 1);
      return gen(...args);
    }
  };
};
const filter = function (gen, pred) {
  const filterGen = function (...args) {
    const val = generator(...args);
    if (assertBool(and(not($eq(val, undefined)), not(pred(val))))) {
      return filterGen(...args);
    }
    return val;
  };
  return filterGen;
};
const concat = function (...gens) {
  const next = element(gens);
  const gen = next();
  const concatGen = function (...args) {
    if (assertBool(not($eq(gen, undefined)))) {
      const val = gen(...args);
      if (assertBool($eq(val, undefined))) {
        gen = next();
        return concatGen(...args);
      }
      return val;
    }
  };
};
const join = function (f, ...gens) {
  return function () {
    const results = gens["map"](function (gen$exclam) {
      return gen$exclam();
    });
    if (
      assertBool(
        not(
          results["every"](function (result$exclam) {
            return $eq(result$exclam, undefined);
          })
        )
      )
    ) {
      return f(...results);
    }
    return undefined;
  };
};
const map = function (f, arr) {
  return join(f, element(arr));
};
const struct = function (...names) {
  return function (...vals) {
    const obj = Object["create"](null);
    names["forEach"](function (name, index) {
      obj[name] = vals[index];
    });
    return obj;
  };
};
module.exports = stone({
  ["constant"]: constant,
  ["integer"]: integer,
  ["element"]: element,
  ["property"]: property,
  ["collect"]: collect,
  ["repeat"]: repeat,
  ["harvest"]: harvest,
  ["limit"]: limit,
  ["filter"]: filter,
  ["concat"]: concat,
  ["join"]: join,
  ["map"]: map,
  ["struct"]: struct,
});
