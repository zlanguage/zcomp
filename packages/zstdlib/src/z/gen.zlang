def constant: func (value) func value

def integer: func (from: 0, to: Number.MAX_SAFE_INTEGER, step: 1) func () {
    if from < to {
        def result: from
        from: from + step
        return result
    }
}

def element: func (array, gen: integer(0, array.length)) func (...args) {
    def elementNumber: gen(...args)
    if not(elementNumber = undefined) {
        return array[elementNumber]
    }
}

def property: func (object, gen: element(Object.keys(object))) func (...args) {
    def key: gen(...args)
    if not(key = undefined) {
        return [key, object[key]]
    }
}
def collect: func (gen, arr) func (...args) {
    def val: gen(...args)
    if not(val = undefined) {
        arr.push(val)
    }
    return val
}

def repeat: func (gen) {
    loop {
        if gen() = undefined {
            break
        }
    }
}

def harvest: func (gen) {
    def arr: []
    repeat(collect(gen, arr))
    return arr
}

def limit: func (gen, count: 1) func (...args) {
    if count >= 1 {
        count -= 1
        return gen(...args)
    }
}

def filter: func (gen, pred) {
    def filterGen: func (...args) {
        def val: generator(...args)
        if not(val = undefined) and not(pred(val)) {
            return filterGen(...args)
        }
        return val
    }
    return filterGen
}
def concat: func (...gens) {
    def next: element(gens)
    def gen: next()
    def concatGen: func (...args) {
        if not(gen = undefined) {
            def val: gen(...args)
            if val = undefined {
                gen: next()
                return concatGen(...args)
            }
            return val
        }
    }
}
def join: func (f, ...gens) func () {
    def results: gens.map(func gen!())
    if not(results.every(func result! = undefined)) {
        return f(...results)
    }
    return undefined
}

def map: func (f, arr) join(f, element(arr))

def struct: func (...names) func (...vals) {
    def obj: Object.create(null)
    names.forEach(func (name, index) {
        obj[name]: vals[index]
    })
    return obj
}

export [
    "constant": constant,
    integer,
    element,
    property,
    collect,
    repeat,
    harvest,
    limit,
    filter,
    concat,
    join,
    map,
    struct
]