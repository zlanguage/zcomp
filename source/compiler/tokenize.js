const symbolMap = {
    "+": "$plus",
    "-": "$minus",
    "*": "$star",
    "/": "$slash",
    "^": "$carot",
    "?": "$question",
    "=": "$eq",
    "<": "$lt",
    ">": "$gt",
    "\\": "$backslash",
    "&": "$and",
    "|": "$or",
    "%": "$percent",
    "'": "$quote",
    "!": "$exclam"
};
const symbolRegExps = Object.keys(symbolMap).map(x => {
    const res = new RegExp(`\\${x}`, "g");
    res.str = symbolMap[x];
    return res;
});
//@ts-check
const unicodeEscapementRegExp = /\\u\{([0-9A-F]{4,6})\}/g;
const newLineRegExp = /\n|\r\n?/;
/*
Capturing Group:
[1] Whitespace
[2] Comment
[3] String
[4] Keyword
[5] Name
[6] Number
[7] Punctuator
[8] RegExp
*/
const tokenRegExp = /(\u0020+|\t+)|(#.*)|("(?:[^"\\]|\\(?:[nrt"\\]|u\{[0-9A-F]{4,6}\}))*")|\b(let|loop|if|else|func|break|import|export|match|return|def|try|on|settle|raise|importstd|meta|enter|exit|operator|hoist|go|get|enum|where|derives|static|macro|include|includestd)\b|([A-Za-z_+\-/*%&|?^=<>'!][A-Za-z_0-9+\-/*%&|?^<>='!]*)|((?:0[box])?-?\d[\d_]*(?:\.[\d_]+)?(?:e\-?[\d_]+)?[a-z]*)|(\.{2,3}|~{|}~|[~@$(),.{}\[\]:])|(`.+?`[gimsuy]*)/y;
/**
 * Create a token generator for a specific source.
 * @param {string | Array<string>} source If string is provided, string is split along newlines. If array is provided, array is used as the array of lines.
 * @param {boolean} comment Should comments be tokenized and returned from the generator?
 * @returns {() => void | {id: string, lineNumber: number, columnNumber: number, string ?: string, columnTo ?: number, readonly ?: boolean, alphanumeric ?: boolean, number ?: number, source ?: string}} A function that generates the next token.
 */
function tokenize(source, comment = false) {
    // Handle multiline comments
    source = source.replace(new RegExp("/\\*[^*]*\\*+(?:[^/*][^*]*\\*+)*/", "g"), "");
    // Configure source
    const lines = (
        Array.isArray(source) ?
        source :
        source.split(newLineRegExp)
    );
    let lineNumber = 0;
    // Refinements are not mangled
    let dotLast = false;
    let insideTemplate = false;
    let template = [];
    let linesDone = [];
    let line = lines[0];
    // Keep trackc of where the token is on the line
    tokenRegExp.lastIndex = 0;
    return function tokenGenerator() {
        // Are we done yet?
        if (line === undefined) {
            return;
        }
        let columnNumber = tokenRegExp.lastIndex;
        // Do we need to go to the next line?
        if (columnNumber >= line.length) {
            tokenRegExp.lastIndex = 0;
            lineNumber += 1;
            line = lines[lineNumber];
            // Get the next token.
            return (
                line === undefined ?
                undefined :
                tokenGenerator()
            );
        }
        console.log
            // Figure out the current token.
        let captives = tokenRegExp.exec(line);
        if (insideTemplate && captives[7] !== "}~" && !linesDone.includes(lineNumber)) {
            template.push(line);
            linesDone.push(lineNumber)
            return tokenGenerator();
        }
        if (insideTemplate && captives[7] !== "}~") {
            return tokenGenerator();
        }
        // If no tokens were matched, it's an error.
        if (!captives) {
            let res = {
                id: "(error)",
                lineNumber,
                columnNumber,
                columnTo: columnNumber,
                string: line.slice(columnNumber)
            }
            line = undefined;
            return res;
        }
        // Whitespace matched
        if (captives[1]) {
            dotLast = false;
            // Ignore whitespace - Z has flexible whitespace rules.
            return tokenGenerator();
        }
        const columnTo = tokenRegExp.lastIndex;
        // Comment matched
        if (captives[2]) {
            dotLast = false;
            // Ignore empty comments
            return (
                comment ? {
                    id: "(comment)",
                    comment: captives[2],
                    lineNumber,
                    columnNumber,
                    columnTo
                } :
                tokenGenerator()
            )
        }
        // String Matched
        if (captives[3]) {
            dotLast = false;
            return {
                id: "(string)",
                readonly: true,
                string: JSON.parse(captives[3].replace( // Handle unicode escapement.
                    unicodeEscapementRegExp,
                    function(ignore, code) {
                        //@ts-ignore
                        return String.fromCodePoint(parseInt(code, 16));
                    }
                )),
                lineNumber,
                columnNumber,
                columnTo
            };
        }
        // Keyword Matched
        if (captives[4]) {
            dotLast = false;
            return {
                id: "(keyword)",
                string: captives[4],
                lineNumber,
                columnNumber,
                columnTo
            };
        }
        // Name Matched
        if (captives[5]) {
            let res = captives[5];
            if (!dotLast) { // If a . wasn't last (the identifier isn't part of a refinement/property access)
                // Mangle it
                symbolRegExps.forEach(regexp => {
                    res = res.replace(regexp, regexp.str);
                });
            }
            // Allow negative numbers.
            if (/\$minus\d+/.test(res)) {
                res = res.replace("$minus", "-")
                return {
                    id: "(number)",
                    readonly: true,
                    number: Number(res.replace(/(?:[^\d])([^0])[a-z]+/g, "$1").replace(/[a-z]$/, "").replace(/_/g, "")),
                    string: res,
                    lineNumber,
                    columnNumber,
                    columnTo
                };
            }
            dotLast = false;
            return {
                id: res,
                source: captives[5],
                alphanumeric: true,
                lineNumber,
                columnNumber,
                columnTo
            };
        }
        // Number Matched
        if (captives[6]) {
            dotLast = false;
            return {
                id: "(number)",
                readonly: true,
                number: Number(captives[6].replace(/(?:[^\d])([^0])[a-z]+/g, "$1").replace(/[a-z]$/, "").replace(/_/g, "")), // Make the number JS-Compatable for parsing
                string: captives[6],
                lineNumber,
                columnNumber,
                columnTo
            };
        }
        // Punctuator Matched
        if (captives[7]) {
            if (captives[7] === ".") {
                dotLast = true; // This means a refinement is coming. The next alphanumeric identifier will not be mangled.
            } else {
                dotLast = false;
            }
            if (captives[7] === "~{") {
                insideTemplate = true;
                return tokenGenerator();
            }
            if (captives[7] === "}~") {
                insideTemplate = false;
                const temp = template;
                template = [];
                linesDone = [];
                return {
                    id: "(template)",
                    string: temp.slice(0, -1).map(x => x.trim()).join("\n"),
                    lineNumber,
                    columnNumber,
                    columnTo
                }
            }
            return {
                id: captives[7],
                lineNumber,
                columnNumber,
                columnTo
            }
        }
        if (captives[8]) {
            return {
                id: "(regexp)",
                string: captives[8].replace(/^`|`(.+)$/g, "").replace(/`$/, ""),
                flags: captives[8].split("`")[2],
                lineNumber,
                columnNumber,
                columnTo
            }
        }
    }
}
module.exports = Object.freeze(tokenize);