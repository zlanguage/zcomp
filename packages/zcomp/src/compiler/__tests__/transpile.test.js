import tokenize from "../tokenize";
import parse from "../parse";
import gen from "../gen";

const evalTests = {
  "math and string manipulations": {
    "3 + 3": 6,
    "2 - 2": 0,
    "2 - 9": -7,
    "3 * 7": 21,
    "6 / 2": 3,
    "3 + 2 * 7": 17,
    "3 / 3 - 2": -1,
    "0.5 + 0.25": 0.75,
    "0.125 - 0.25 * 0.5": 0,
    "0.125 / 4 * 10": 0.3125,
    "-3.2 + -3": -6.2,
    "2 ^ 3": 8,
    "2 pow 3": 8,
    "2 ^ 3 ^ 3": 512,
    "2 pow 3 pow 3": 134217728,
    "1 pow 0": 1,
    "2 * 3 ^ 3 + 2": 56,
    "5 % 2": 1,
    "5 % 2 + 1": 2,
    "[1, 2, 3] ++ [4, 5, 6]": [1, 2, 3, 4, 5, 6],
    '"Hello" ++ " World"': "Hello World",
    '"[" ++ [1, 2, 3] ++ "]"': "[1,2,3]",
  },
  "relational and boolean logic operators": {
    "true and false": false,
    "true or false": true,
    "true and false or false": false,
    "3 = 3": true,
    "0.1 + 0.2 = 0.3": true,
    "2 = 3": false,
    '"Hello" = "Hello"': true,
    '"Hello" = "World"': false,
    'not("Hello" = "World")': true,
    "[1, 2, 3] = (1, 2, 3)": true,
    "[[1], [2], [3]] = ([1], [2], [3])": true,
    "[1, 2, 4] = {1, 2, 3}": false,
    '["x": 3, "y": 4] = ["x": 3, "y": 4]': true,
    '["x": 3, "y": 5] = ["x": 3, "y": 4]': false,
    "3 > 2": true,
    "2 > 3": false,
    "2 < 3": true,
    "3 < 2": false,
    "3 >= 3": true,
    "3 >= 2": true,
    "3 >= 4": false,
    "3 <= 3": true,
    "3 <= 4": true,
    "3 <= 2": false,
    "copy(if (true) 3 else 5)": 3,
    "copy(if (false) 3 else 5)": 5,
    "copy(if (true) 3 else if (false) 7 else 4)": 3,
    "copy(if (false) 3 else if (true) 7 else 4)": 7,
    "copy(if (false) 3 else if (false) 7 else 4)": 4,
  },
  "refinements, subscripts, and invocations": {
    "copy(func(){}())": undefined,
    "copy(console.log)": console.log,
    'copy(console["log"])': console.log,
    "copy(console.log.call)": console.log.call,
    'copy(console["log"].call)': console.log.call,
    "copy(console..log..call)": console.log.call,
    "copy(console..foo..call)": undefined,
    'copy("Hello").slice(0, -1)': "Hell",
    "copy(undefined..x..y)": undefined,
    '"Hello"..slice(0, -1)': "Hell",
  },
  functions: {
    "copy(func x! + 2)(2)": 4,
    "copy(func () @@iterator)()": Symbol.iterator,
    [`copy(func (x number!, y number!) number! {
            return x + y
        })(3, 4)`]: 7,
  },
  enums: {
    [`enum Point(x: number!, y: number!)
        Point(3, 4).x
        `]: 3,
    [`enum Point(x: number!, y: number!)
        Point(y: 3, x: 4).y
        `]: 3,
    [`importstd traits
        def { Show, Enum }: traits
        enum Color {
            Red,
            Orange,
            Yellow,
            Green,
            Blue,
            Purple
        } derives (Show, Enum)
        Red().succ().toString()`]: "Orange()",
  },
  "loop expressions": {
    "[] ++ loop(x <- 1...10) x": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "[] ++ loop(x <- 1 to 10, if x % 2 = 0) x * 2": [4, 8, 12, 16, 20],
    "[] ++ loop(x <- 1 to 10, y: x * 2, if y % 4 = 0) x": [2, 4, 6, 8, 10],
    "[] ++ loop(x <- 1 to 3, if x % 2 = 0, y <- 1 to 3, z: x * y) z": [2, 4, 6],
  },
  "pattern matching": {
    [`match 3 {
            3 => "yes",
            _ => "no"
        }`]: "yes",
    [`match 4 {
            3 => "yes",
            _ => "no"
        }`]: "no",
    [`match [1, 2, 3] {
            (number!n, ...xs) => xs,
            _ => "no"
        }`]: [2, 3],
    [`match 3 {
            number! => "A number.",
            string! => "A string.",
            array! => "An array",
            _ => "Something else."
        }`]: "A number.",
    [`match "hola" {
            number! => "A number.",
            string! => "A string.",
            array! => "An array",
            _ => "Something else."
        }`]: "A string.",
    [`match [1, 2, 3] {
            number! => "A number.",
            string! => "A string.",
            array! => "An array",
            _ => "Something else."
        }`]: "An array",
    [`match \`a\` {
            number! => "A number.",
            string! => "A string.",
            array! => "An array",
            _ => "Something else."
        }`]: "Something else.",
    [`match ["x": 3, "y": 4] {
            { x: number!, y: number! } => "A point-like object.",
            _ => "no"
        }`]: "A point-like object.",
  },
  macros: {
    [`macro $hello () {
            return ~{
                "Hey"
            }~
        }
        copy($hello)`]: "Hey",
    [`macro $join(~l:expr, ~r:expr) {
            return ~{
                {{~l}} ++ " " ++ {{~r}}
            }~
        }
        copy($join "Hello" "World")`]: "Hello World",
    [`macro $proc(~body:block) {
            return ~{
                func () {
                    {{~body}}
                }
            }~
        }
        copy($proc {
            return 3
        })()`]: 3,
    [`macro operator ->(~param:expr, ~body:expr) {
            return ~{
                func ({{~param}}) {
                    return {{~body}}
                }
            }~
        }
        copy(x -> x * 2)(5)`]: 10,
    [`macro $do (~body:block, while, ~cond:expr) {
            return ~{
               loop {
                   {{~body}}
                   if not({{~cond}}) {
                       break
                   }
               }
           }~
       }

       macro $do (~body:block, until, ~cond:expr) {
            return ~{
               loop {
                   {{~body}}
                   if {{~cond}} {
                       break
                   }
               }
           }~
       }

       macro $do (~body:block, unless, ~cond:expr) {
            return ~{
             if not({{~cond}}) {
               {{~body}}
             }
           }~
       }`]: "use strict",
    [`macro $switch (...{case, ~body:block},) {

       }`]: "use strict",
  },
};

const transpileTests = {
  "variable declarations": [
    "let x: 0",
    "let x: 0, y: 0",
    "def x: 0",
    "def x: 0, y: 0",
    "hoist x: 0",
    "hoist x: 0, y: 0",
  ],
  modules: [
    "import foo",
    'import foo: "./foo"',
    "import ramda.src.map",
    "importstd gr",
    "export fooey",
  ],
  "$Z Macro": ["log($Z)"],
  include: ['includestd "imperative"'],
  "control flow": [
    "if foo { log(bar) }",
    "log(bar) if foo",
    "if foo { log(bar) } else if fooey { log(baree) } else { log(foobar) }",
    'loop { log("YAY") }',
  ],
  "error handling": [
    `try {
            raise "FOOEY"
          } on err {
            settle err
          }`,
  ],
  goroutines: [
    `copy(go func () {
            get fooey()
          })`,
    `go {
            get fooey()
        }`,
    "get foo",
  ],
  comments: ["# Hola"],
  enums: [
    `enum Foo {
            Bar(x: number!),
            Baz(y: _!, z: number!)
          } derives (Nada) where {
            foobar () {}
          } `,
    `enum Point(x: number!, y: number!) where {
            nada () {}
          }`,
  ],
  "advanced function & features": [
    "x number!: 3",
    `func (x number!) number! {
        return x
      }`,
    'func () { exit { log("Hello World") } }',
    "Math.pow(@, @)",
    ".val",
  ],
  metadata: ["operator +: 1000"],
};

function evalZ(z) {
  return eval(gen(parse(tokenize(z))));
}

function transpileZ(z) {
  return gen(parse(tokenize(z)), false);
}

describe("Evaluation tests", () => {
  Object.entries(evalTests).forEach(([testName, tests]) => {
    describe(testName, () => {
      Object.entries(tests).forEach(([expr, res]) => {
        it(`should evaluate ${expr} as ${
          res == null ? res : res.toString()
        }`, () => {
          expect(evalZ(expr)).toStrictEqual(res);
        });
      });
    });
  });
});

describe("Transpiling snapshot tests", () => {
  Object.entries(transpileTests).forEach(([testName, tests]) => {
    describe(`(${testName})`, () => {
      tests.forEach((test) => {
        it(`should transpile ${test} to the snapshot`, () => {
          expect(
            transpileZ(test)
              .split("\n")
              .map((x) => x.trim())
              .join("")
          ).toMatchSnapshot();
        });
      });
    });
  });
});
