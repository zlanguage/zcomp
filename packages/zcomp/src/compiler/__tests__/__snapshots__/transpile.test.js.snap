// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpiling snapshot tests ($Z Macro) should transpile log($Z) to the snapshot 1`] = `"log($Z);"`;

exports[`Transpiling snapshot tests (advanced function & features) should transpile .val to the snapshot 1`] = `"function (obj) {return obj[\\"val\\"];};"`;

exports[`Transpiling snapshot tests (advanced function & features) should transpile Math.pow(@, @) to the snapshot 1`] = `"curry(function ($, $$) {return Math[\\"pow\\"]($, $$);});"`;

exports[`Transpiling snapshot tests (advanced function & features) should transpile func () { exit { log("Hello World") } } to the snapshot 1`] = `"function (){ try {  {} } finally { if (!(log(\\"Hello World\\"))) { throw new Error(\\"Exit failed\\") } } };"`;

exports[`Transpiling snapshot tests (advanced function & features) should transpile func (x number!) number! {
        return x
      } to the snapshot 1`] = `"function (x) {if (!($eq(typeOf(x), \\"number\\"))) { throw new Error(\\"Enter failed\\") }return assertType(\\"number\\", x);};"`;

exports[`Transpiling snapshot tests (advanced function & features) should transpile x number!: 3 to the snapshot 1`] = `"x = assertType(\\"number\\", 3);"`;

exports[`Transpiling snapshot tests (comments) should transpile # Hola to the snapshot 1`] = `""`;

exports[`Transpiling snapshot tests (control flow) should transpile if foo { log(bar) } else if fooey { log(baree) } else { log(foobar) } to the snapshot 1`] = `"if (assertBool(foo)) {log(bar);} else {if (assertBool(fooey)) {log(baree);} else {log(foobar);}}"`;

exports[`Transpiling snapshot tests (control flow) should transpile if foo { log(bar) } to the snapshot 1`] = `"if (assertBool(foo)) {log(bar);}"`;

exports[`Transpiling snapshot tests (control flow) should transpile log(bar) if foo to the snapshot 1`] = `"if (assertBool(foo)) {log(bar);}"`;

exports[`Transpiling snapshot tests (control flow) should transpile loop { log("YAY") } to the snapshot 1`] = `"while (true) {log(\\"YAY\\");}"`;

exports[`Transpiling snapshot tests (enums) should transpile enum Foo {
            Bar(x: number!),
            Baz(y: _!, z: number!)
          } derives (Nada) where {
            foobar () {}
          }  to the snapshot 1`] = `"function Bar(x) {if($eq(Object.keys((x == null) ? { [Symbol()]: 0 } : x).sort(), [\\"x\\"].sort())) {({ x } = x);}if (typeOf(x) !== \\"number\\") {throw new Error(\\"Foo.Bar.x must be of type number. However, you passed \\" + x + \\" to Foo.Bar which is not of type number.\\");}return Nada({type() { return \\"Foo\\"; },get constructor() { return Bar; },get parent() { return Foo; },get fields() { return [\\"x\\"]; },get x(){ return x; },\\"=\\"(other) {return other.constructor === Bar && $eq(x, other.x);}});}Bar.extract = function (val) {if (val.constructor === Bar) {return [val.x];}return undefined;};function Baz(y, z) {if($eq(Object.keys((y == null) ? { [Symbol()]: 0 } : y).sort(), [\\"y\\", \\"z\\"].sort())) {({ y, z } = y);}if (typeOf(z) !== \\"number\\") {throw new Error(\\"Foo.Baz.z must be of type number. However, you passed \\" + z + \\" to Foo.Baz which is not of type number.\\");}return Nada({type() { return \\"Foo\\"; },get constructor() { return Baz; },get parent() { return Foo; },get fields() { return [\\"y\\", \\"z\\"]; },get y(){ return y; },get z(){ return z; },\\"=\\"(other) {return other.constructor === Baz && $eq(y, other.y) && $eq(z, other.z);}});}Baz.extract = function (val) {if (val.constructor === Baz) {return [val.y, val.z];}return undefined;};var Foo = {order: [Bar, Baz],Bar,Baz};Foo.foobar = function () {};"`;

exports[`Transpiling snapshot tests (enums) should transpile enum Point(x: number!, y: number!) where {
            nada () {}
          } to the snapshot 1`] = `"function Point(x, y) {if($eq(Object.keys((x == null) ? { [Symbol()]: 0 } : x).sort(), [\\"x\\", \\"y\\"].sort())) {({ x, y } = x);}if (typeOf(x) !== \\"number\\") {throw new Error(\\"Point.Point.x must be of type number. However, you passed \\" + x + \\" to Point.Point which is not of type number.\\");}if (typeOf(y) !== \\"number\\") {throw new Error(\\"Point.Point.y must be of type number. However, you passed \\" + y + \\" to Point.Point which is not of type number.\\");}return {type() { return \\"Point\\"; },get constructor() { return Point; },get parent() { return Point; },get fields() { return [\\"x\\", \\"y\\"]; },get x(){ return x; },get y(){ return y; },\\"=\\"(other) {return other.constructor === Point && $eq(x, other.x) && $eq(y, other.y);}};}Point.extract = function (val) {if (val.constructor === Point) {return [val.x, val.y];}return undefined;};Point.order = [Point];Point.nada = function () {};"`;

exports[`Transpiling snapshot tests (error handling) should transpile try {
            raise "FOOEY"
          } on err {
            settle err
          } to the snapshot 1`] = `"try {throw new Error(\\"FOOEY\\");} catch (err) {err[\\"settled\\"] = true;if (assertBool($eq(err[\\"settled\\"], undefined))) {throw new Error(\\"Error err not settled.\\")}}"`;

exports[`Transpiling snapshot tests (goroutines) should transpile copy(go func () {
            get fooey()
          }) to the snapshot 1`] = `"copy(async function () {await fooey()._from();});"`;

exports[`Transpiling snapshot tests (goroutines) should transpile get foo to the snapshot 1`] = `"var $main = async function () {await foo._from();};$main();"`;

exports[`Transpiling snapshot tests (goroutines) should transpile go {
            get fooey()
        } to the snapshot 1`] = `"(async function () {await fooey()._from();})();"`;

exports[`Transpiling snapshot tests (include) should transpile includestd "imperative" to the snapshot 1`] = `""`;

exports[`Transpiling snapshot tests (metadata) should transpile operator +: 1000 to the snapshot 1`] = `"/* operator $plus = 1000 */"`;

exports[`Transpiling snapshot tests (modules) should transpile export fooey to the snapshot 1`] = `"module.exports = stone(fooey);"`;

exports[`Transpiling snapshot tests (modules) should transpile import foo to the snapshot 1`] = `"var foo = stone(require(\\"foo\\"));"`;

exports[`Transpiling snapshot tests (modules) should transpile import foo: "./foo" to the snapshot 1`] = `"var foo = stone(require(\\"./foo\\"));"`;

exports[`Transpiling snapshot tests (modules) should transpile import ramda.src.map to the snapshot 1`] = `"var map = stone(require(\\"ramda/src/map\\"));"`;

exports[`Transpiling snapshot tests (modules) should transpile importstd gr to the snapshot 1`] = `"var gr = stone(require(\\"@zlanguage/zstdlib/src/js/gr\\"));"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile def x: 0 to the snapshot 1`] = `"var x = 0;"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile def x: 0, y: 0 to the snapshot 1`] = `"var x = 0, y = 0;"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile hoist x: 0 to the snapshot 1`] = `"var x = 0;"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile hoist x: 0, y: 0 to the snapshot 1`] = `"var x = 0, y = 0;"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile let x: 0 to the snapshot 1`] = `"let x = 0;"`;

exports[`Transpiling snapshot tests (variable declarations) should transpile let x: 0, y: 0 to the snapshot 1`] = `"let x = 0, y = 0;"`;
